### DISCUSSION TITLE
【React知识点】

### DISCUSSION THREAD

**Discussion by swustzzh** [swustzzh/coder_books#22] [General]
- useEffect第二个参数中的依赖是基于浅对比处理的
- React Fiber
  - 16之前，协调剁成是通过递归调用render方法来实现，问题在于递归很深的话，会占用较长时间，阻塞主线程，导致用户界面不流畅；其次递归无法中断
  - Fiber讲协调过程改为可中断和可恢复的方式，更好的控制任务的优先级，允许执行过程中的暂停、终止和恢复。
    具体来说，Fiber架构的特点包括：
    1. 异步可中断：Fiber使得React的渲染过程变为可中断的，可以根据任务的优先级来中断当前操作，并在空闲时恢复。这样可以更好地响应用户的交互操作，提升用户体验。
    2. 增量更新：Fiber允许渲染过程的增量更新，即将整个渲染任务拆分为多个小任务，通过时间切片的方式逐步完成。这样可以避免长时间的阻塞，使得应用程序更加流畅。
    3. 优先级调度：Fiber引入了任务优先级的概念，React可以根据任务的优先级来决定是否中断当前任务，并立即执行高优先级的任务。这样可以更好地控制任务的执行顺序，提高系统的响应能力。
    4. 错误边界：Fiber提供了错误边界（Error Boundary）的机制，可以捕获和处理组件内部发生的错误，防止整个应用程序崩溃。
  - Fiber架构中的任务不是真正的线程或进程，而是通过JS的事件循环机制来实现
  - > 在执行渲染任务时，React会首先创建一个Fiber树来表示整个UI。然后，React通过遍历Fiber树来执行不同的渲染任务，并将这些任务分成多个小的任务单元，称为“工作单元”（work unit）或“工作项”。
     > 每个工作单元都是一个JavaScript对象，包含一个执行特定任务的函数，以及其他必要的参数。当React开始执行渲染任务时，它会从Fiber树的根节点开始，依次遍历每个Fiber节点，判断该节点需要执行哪些工作单元。React会将这些工作单元放到一个叫做“工作队列”（work queue）的结构中，按照优先级顺序排列。
     > React使用Fiber节点的effect list和effect tag来记录终端任务并恢复。Fiber节点的effect list是一个链表结构，用于记录该节点及其子节点产生的DOM操作，如插入、更新和删除等。effect tag是一个标志位，用于标识该节点产生的DOM操作类型。
     > 当React执行任务时，如果需要进行DOM操作，会创建一个包含该操作信息的effect对象，并添加到当前Fiber节点的effect list中。这个effect对象包含了一些描述该操作的信息，比如操作类型、DOM节点等。当React在下一次渲染时，可以遍历每个Fiber节点的effect list，并根据其中的信息执行对应的DOM操作，从而恢复之前中断的工作进度。

**Comment by swlws** [comment]
[智能组件和木偶组件](https://www.cnblogs.com/rik28/p/8676151.html)

智能组件：仅包含数据和行为，不涉及UI
木偶组件：仅做UI渲染
